### ISSUE TITLE
quarkus-cache: "keyGenerator" destroyed, even if it is annotated with "Singleton"

### ISSUE THREAD

**Issue by wglanzer** [quarkusio/quarkus#38394]
### Describe the bug

The key generator seems to be destroyed, even if it is marked as `Singleton`.
This seems a bit confusing, because I thought, that the key generator is only instantiated once - as the annotation `Singleton` says.

A simple CacheKeyGenerator instance:
```java
@Singleton // We only want a single instance here
public class RequestContextKeyGenerator implements CacheKeyGenerator
{
  public RequestContextKeyGenerator()
  {
    System.out.println("created"); // <-- Gets logged on EVERY calculateSourceRoot() invocation
  }

  @Override
  public Object generate(Method method, Object... methodParams)
  {
    // do some fancy key logic here
  }
}
```

CacheKeyGenerator used in a demo method:
```java
@CacheResult(cacheName = "sourceRoots", keyGenerator = RequestContextKeyGenerator.class)
protected Path calculateSourceRoot(@NonNull Path pPath)
{
  // do some calculations here
}
```

Let's say I call the `calculateSourceRoot` method 5 times; the key generator gets instantiated 5 times too.

### Expected behavior

I think it is not necessary to destroy the bean after key creation - the key generator bean should be reused, especially if it is marked as `Singleton`.

### Actual behavior

The key generator gets destroyed after every cache-method invocation.
It seems, that `keyGenerator.destroy(keyGen);` in io.quarkus.cache.runtime.CacheInterceptor#generateKey() should not be invoked, if the bean is a `Singleton` instance.
Then the generator would be reused properly.

**Comment by geoand** [comment]
This is happening because of https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java#L165. 

@mkouba what is the proper pattern to use here?

**Comment by mkouba** [comment]
> This is happening because of
> 
> https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java#L165
> .
> 
> @mkouba what is the proper pattern to use here?

It surely does not make sense for a `@Singleton` implementation that is designed to be shared... although it's legal. And since the [docs](https://quarkus.io/guides/cache#generating-a-cache-key-with-cachekeygenerator) mention that _"All CDI scopes supported by Quarkus can be used on a key generator."_ the logic should be fixed.

It would make sense to destroy `@Dependent` beans only (which cannot be shared anyway).

The easies way to do this would be something like:
```java
Instance<T> keyGenInstance = keyGenerator.select(keyGeneratorClass);
if (keyGenInstance.isResolvable()) {
   LOGGER.tracef("Using cache key generator bean from Arc [class=%s]", keyGeneratorClass.getName());
   Handle<T> keyGen = keyGenInstance.getHandle();
   try {
      return keyGen.get().generate(method, methodParameterValues);
   } finally {
      // Quarkus-specific
      keyGen.close();
      // Spec-compliant
      InjectableBean<T> bean = keyGen.getBean();
      if (bean == null || Dependent.class.equals(bean.getScope())) {
         destroy();
       }
   }
}
```

OT: It's not exactly clear to name a `@Singleton` implementation `RequestContextKeyGenerator` :D

**Comment by geoand** [comment]
Thanks a lot @mkouba!

Would you like to go ahead and provide a PR?

**Comment by mkouba** [comment]
> Thanks a lot @mkouba!
> 
> Would you like to go ahead and provide a PR?

For sure! I'll add @gwenneg  for the review.

**Comment by geoand** [comment]
Thanks!

**Comment by wglanzer** [comment]
> OT: It's not exactly clear to name a `@Singleton` implementation `RequestContextKeyGenerator` :D

@mkouba First of all, thank you for your really fast fix! 
You're completely right, that the naming seems a bit off here. Initially the `Singleton` annotation was not present, because I noticed, that the class gets created every time. So, in production, I removed the annotation to skip the whole CDI flow and improve the performance overall.
After this "fix" I created this issue and .. I completely forgot about the class name here :D
